coffee/expander.py

```diff
@@ -0,0 +1,244 @@
+
+        return {exp: hoisted_exp}
+
+    def _build(self, exp, grp):
+        """Create a node for the expansion and keep track of it."""
+        expansion = Prod(exp, dcopy(grp))
+        # Track the new expansion
+        self.expansions.append(expansion)
+        # Untrack any expansions occured in children nodes
+        if grp in self.expansions:
+            self.expansions.remove(grp)
+        return expansion
+
+    def _expand(self, node, parent):
+        if isinstance(node, Symbol):
+            return ([node], self.EXPAND) if self.should_expand(node) \
+                else ([node], self.GROUP)
+
+        elif isinstance(node, Par):
+            return self._expand(node.child, node)
+
+        elif isinstance(node, (Div, FunCall)):
+            # Try to expand /within/ the children, but then return saying "I'm not
+            # expandable any further"
+            for n in node.children:
+                self._expand(n, node)
+            return ([node], self.GROUP)
+
+        elif isinstance(node, Prod):
+            l_exps, l_type = self._expand(node.left, node)
+            r_exps, r_type = self._expand(node.right, node)
+            if l_type == self.GROUP and r_type == self.GROUP:
+                return ([node], self.GROUP)
+            # At least one child is expandable (marked as EXPAND), whereas the
+            # other could either be expandable as well or groupable (marked
+            # as GROUP): so we can perform the expansion
+            groupable = l_exps if l_type == self.GROUP else r_exps
+            expandable = r_exps if l_type == self.GROUP else l_exps
+            to_replace = OrderedDict()
+            for exp, grp in itertools.product(expandable, groupable):
+                expansion = self._build(exp, grp)
+                to_replace.setdefault(exp, []).append(expansion)
+            ast_replace(node, {k: ast_make_expr(Sum, v) for k, v in to_replace.items()},
+                        mode='symbol')
+            # Update the parent node, since an expression has just been expanded
+            expanded = node.right if l_type == self.GROUP else node.left
+            parent.children[parent.children.index(node)] = expanded
+            return (list(flatten(to_replace.values())) or [expanded], self.EXPAND)
+
+        elif isinstance(node, (Sum, Sub)):
+            l_exps, l_type = self._expand(node.left, node)
+            r_exps, r_type = self._expand(node.right, node)
+            if l_type == self.EXPAND and r_type == self.EXPAND and isinstance(node, Sum):
+                return (l_exps + r_exps, self.EXPAND)
+            elif l_type == self.EXPAND and r_type == self.EXPAND and isinstance(node, Sub):
+                return (l_exps + [Neg(r) for r in r_exps], self.EXPAND)
+            else:
+                return ([node], self.GROUP)
+
+        else:
+            raise RuntimeError("Expansion error: unknown node: %s" % str(node))
+
+    def expand(self, should_expand, **kwargs):
```

Comment:
just the main expander function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)